File: server/src/database/backup.rs
use chrono::{DateTime, Duration, Utc};
use common::DATE_TIME_FORMAT;
use rocket::tokio::time::sleep;
use rocket::tokio::fs;
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize)]
struct Backup {
    last_back_up: DateTime<Utc>,
}
const BACKUP_PATH: &str = "./data/backup.json";
lazy_static! {
    static ref INTERVAL: Duration = Duration::days(14);
}

pub async fn backup() {
    let now = Utc::now();
    let mut till_update = fs::read_to_string(BACKUP_PATH)
        .await
        .ok()
        .and_then(|contents| serde_json::from_str::<Backup>(&contents).ok())
        .map(|json| json.last_back_up + *INTERVAL - now)
        .unwrap_or_else(|| Duration::zero());
    
    if till_update.num_seconds() < 0{
        // we should have already updated so set to 0
        till_update = Duration::zero();
    }
    
    println!(
        "next backup scheduled for {}",
        (now + till_update).format(DATE_TIME_FORMAT)
    );
    
    sleep(till_update.to_std().unwrap()).await;

    loop {
        let now = Utc::now();
        println!("{} - backing up", now.format(DATE_TIME_FORMAT));

        let result = fs::copy(
            "./data/Ulina.db",
            format!("./data/backups/Ulina_{}.db", now.format(DATE_TIME_FORMAT)),
        )
        .await;

        match result {
            Ok(_) => println!("successful backup"),
            Err(e) => println!("error backing up database: {}", e.to_string()),
        };

        let result = fs::write(
            BACKUP_PATH,
            serde_json::to_string(&Backup { last_back_up: now }).unwrap(),
        )
        .await;

        match result {
            Err(e) => println!("error saving backup date: {}", e.to_string()),
            _ => {}
        }

        println!(
            "next backup scheduled for {}",
            (now + *INTERVAL).format(DATE_TIME_FORMAT)
        );

        sleep(INTERVAL.to_std().unwrap()).await;
    }
}

File: server/backup.ts
import * as fs from "fs";
import {formatDateForFile} from "./utilities";

const backUpFilePath = "./data/backups/backup.json";
const interval = 86400000 * 7;

const start = () => {
    const now = Date.now();
    const lastBackUp = JSON.parse(fs.readFileSync(backUpFilePath, "utf-8")).lastBackUp;
    const tillNextBackUp = lastBackUp === null ? 0 : Math.max(0, lastBackUp - now + interval);
    console.log(`next backup scheduled for ${new Date(now + tillNextBackUp)}`);
    setTimeout(backUp, tillNextBackUp);
}

async function backUp(){
    const now= Date.now();
    console.log(`${new Date(now)} - backing up...`);
    fs.writeFileSync(backUpFilePath, JSON.stringify({lastBackUp: now}));
    fs.copyFile("./data/Ulina.db", `./data/backups/${formatDateForFile("Ulina", ".db")}`,
        (err) => {
        if (err !== null) console.log(`error backing up database: ${err}`);
        else {
            console.log("successful backup")
        }

        setTimeout(backUp, interval);
        console.log(`next backup scheduled for ${new Date(now + interval)}`);
    });
}

start()